machineconfiguration/v1alpha1/types_machineconfigstate.go

```diff
@@ -0,0 +1,125 @@
+package v1alpha1
+
+import (
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// +genclient
+// +genclient:nonNamespaced
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// MachineConfigState describes the health of the Machines on the system
+// Compatibility level 4: No compatibility is provided, the API can change at any point for any reason. These capabilities should not be used by applications needing long term support.
+// +openshift:compatibility-gen:level=4
+type MachineConfigState struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	// +kubebuilder:validation:Required
+	Spec MachineConfigStateSpec `json:"spec"`
+	// +optional
+	Status MachineConfigStateStatus `json:"status"`
+}
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// MachineConfigStateList describes all of the MachinesStates on the system
+//
+// Compatibility level 4: No compatibility is provided, the API can change at any point for any reason. These capabilities should not be used by applications needing long term support.
+// +openshift:compatibility-gen:level=4
+type MachineConfigStateList struct {
+	metav1.TypeMeta `json:",inline"`
+	metav1.ListMeta `json:"metadata"`
+
+	Items []MachineConfigState `json:"items"`
+}
+
+// MachineConfigStateSpec describes the type of State we are managing
+type MachineConfigStateSpec struct {
+	// configuration describes the object reference information of a MachineConfigState
+	Config MachineConfigStateConfig `json:"configuration"`
+}
+
+// MachineConfigStateStatus holds the reported information on a particular MachineConfigState
+type MachineConfigStateStatus struct {
+	corev1.ObjectReference `json:",inline"`
+	// conditions represents the latest available observations of current state.
+	// +listType=atomic
+	// +optional
+	Conditions []ProgressionCondition `json:"conditions"`
+	// mostRecentError is populated if the State reports an error.
+	MostRecentError string `json:"mostRecentError"`
+	// health reports the overall status of this MachineConfigState Given its Progress
+	Health MachineConfigStateHealthEnumeration `json:"health"`
+	// observedGeneration represents the generation observed by the controller.
+	// +optional
+	ObservedGeneration int64 `json:"observedGeneration,omitempty"`
```

Comment:
So we are on the same page, can you describe in the godoc when this value is updated?

Judging by the shape of this API at the moment. The spec contains a desired version (same as the annotation on the Node) and not a lot else. This should be updated once the MCD has observed the diff between current and desired, done it's pre-flight checks, and then either errored or accepted that it can move to this new desired version. Does that make sense?

I suspect you can't actually do this based on the annotation based API MCD currently uses and this would require MCD to update the MCS directly

Reply:
it is updated in the MCO upon `ApplyStatus` of the next MachineConfigState. but yeah, I can definitely make this instead change when the MCD "accepts" it needs to update :) I have learned that there is a surprising amount of information hidden in the MCO annotations and behavior related to the node.